common/browser_process.go

```diff
@@ -38,23 +39,26 @@ func NewBrowserProcess(
 	ctx context.Context, path string, args, env []string, dataDir *storage.Dir,
 	ctxCancel context.CancelFunc, logger *log.Logger,
 ) (*BrowserProcess, error) {
-	cmd, stdout, procDone, err := execute(ctx, path, args, env, dataDir, logger)
+	procCtx, procCtxCancel := context.WithCancel(ctx)
+	cmd, err := execute(
+		procCtx, procCtxCancel, path, args, env, dataDir, logger)
 	if err != nil {
+		procCtxCancel()
 		return nil, err
 	}
 
-	wsURL, err := parseDevToolsURL(ctx, stdout)
+	wsURL, err := parseDevToolsURL(cmd)
 	if err != nil {
-		return nil, fmt.Errorf("getting DevTools URL: %w", err)
+		procCtxCancel()
+		return nil, err
 	}
 
 	p := BrowserProcess{
-		ctx:                        ctx,
+		ctx:                        procCtx,
 		cancel:                     ctxCancel,
```

Comment:
I didn't think about it much TBH, but looking at it now, we want `BrowserProcess.ctx` to be a new context associated with the process, in case the caller depends on this specific context being done. This doesn't seem to be the case now, but it should be a separate context only associated to the browser process.

And we're setting the parent context's cancel function, because, well, that cancels the parent context, which interrupts the iteration, etc. If this were `procCtxCancel`, then only the process itself would be cancelled.

Reply:
I guess that makes sense 👍